Cross compile groovy and java where appropriate

grumpyjames opened the pull request:
Summary:

groovyc is able to collaborate with javac in order to compile .java
and .groovy files that express circular dependencies. No comment is
made as to whether this is a good idea, but it is now possible with
buck if one wished to.

By default, the inner javac will be parametrised in the same we it
would be as in a java library rule; these parameters can be overridden
in the same way they can be for a java_library rule.

Test plan: CI (there are new cases in GroovyLibraryIntegrationTest)

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -61,8 +69,8 @@
-
-    ProcessBuilder processBuilder = new ProcessBuilder(createCommand());
+    ImmutableList<String> command = createCommand();
+    ProcessBuilder processBuilder = new ProcessBuilder(command);
```
Two things:
* This change looks spurious
* We shouldn't be adding new uses of `ProcessBuilder` in the codebase.  You didn't do this here, but I'm letting you know.  We should be using `ProcessExecutor` or `ListeningProcessExecutor` instead.

LegNeato replied:
Add a lint if that's the case? Not sure how external people would know and this is perfect for lint.

sdwilsh replied:
We totally should, but I'd rather spend that effort on getting Windows CI which would catch this in the first place (`ProcessBuilder` doesn't properly escape the arguments on Windows, so you might be okay, but you might also fail in a non-obvious way.  The other two classes work around that issue.)

grumpyjames replied:
Agreed on both counts!

grumpyjames replied:
I'm going to undo the spurious part in this PR, but I think switching to ProcessExecutor or equivalent should probably be a separate commit.

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -91,14 +98,61 @@ public String getDescription(ExecutionContext context) {
+
+    command.addAll(groovyc.getCommandPrefix(resolver));
+
+    String classpath = Joiner.on(":").join(transform(declaredClasspathEntries, toStringFunction()));
```
Again, this isn't wrong in this diff since you just moved the code, but we shouldn't hardcode `:` here, but instead use `File.pathSeparator` so this works on Windows too.

LegNeato replied:
Add a lint if that's the case? Not sure how external people would know and this is perfect for lint.

sdwilsh replied:
Same as above; Windows CI would have caught this for us, and I'm not sure how to write a lint rule that would be context-aware enough to know when `:` is safe to use and when it isn't.

grumpyjames replied:
Will fix this.

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -91,14 +98,61 @@ public String getDescription(ExecutionContext context) {
+      javacOptions.appendOptionsTo(new OptionsConsumer() {
+        @Override
+        public void addOptionValue(String option, String value) {
+          // javac won't find things compiled with groovyc otherwise.
```
otherwise?  It seems like some text might be missing here or it is misleading me how it is worded.

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -91,14 +98,61 @@ public String getDescription(ExecutionContext context) {
+        @Override
+        public void addOptionValue(String option, String value) {
+          // javac won't find things compiled with groovyc otherwise.
+          // alternatively, we could convert all the paths to be absolute.
```
when you say "all paths", what do you mean?

grumpyjames replied:
So, by default in AbstractJavacOptions, we have this:

    // Set the sourcepath to stop us reading source files out of jars by mistake.
    optionsConsumer.addOptionValue("sourcepath", "");

It turns out that this doesn't play well with cross compilation, to the point where passing empty sourcepath to the 'inner' javac means it cannot find the groovyc compiled classes.

We found two potential paths to avoiding this problem; the simpler one (just ignore the source path option) is presented here. The more complicated solution involves making every path we pass to the compile step absolute; we liked this less.

I'm not particularly happy with either solution and would welcome any further input. I'll go and think about the possible values of source path to see if we always want to drop it, whether we want to log a warning if it's set, etc...

sdwilsh commented on test/com/facebook/buck/jvm/groovy/GroovyLibraryIntegrationTest.java:
```diff
@@ -66,15 +64,38 @@ public void shouldCompileLibraryWithDependencyOnAnother() throws Exception {
+
+  @Test
+  public void javaOptionsArePassedThroughToTheJavacCompiler() throws Exception {
+    // code requires java 7, source set to java 6
```
nit: please use complete sentences, which means you need capital letters and punctuation at the end.

grumpyjames replied:
Understood!

grumpyjames commented:
Thanks for reviewing. I'll take another look at the sourcepath option handling and add further commits here.

grumpyjames commented:
So, I think I addressed everything, apart from switching from ProcessBuilder to ProcessExecutor, which I'll look at in a separate commit.

I had a further look at sourcepath option; it's only ever set from within `AbstractJavacOptions`, and the setting it provides is incompatible with how cross compilation works. Quietly ignoring it sounds like the right thing to do, so I've updated the comment to explain.

sdwilsh commented:
You've got a lint failure that Travis caught:

    /home/travis/build/facebook/buck/src/com/facebook/buck/jvm/groovy/GroovycStep.java:105: Line is longer than 100

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java:
```diff
@@ -99,7 +106,14 @@ public Arg createUnpopulatedConstructorArg() {
+                        resolver,
+                        pathResolver,
+                        args
+                    )),
```
small nit (that I'd love if we could lint for): we keep the closing paren after `args`

grumpyjames replied:
both of them in this case, I guess (from context that looks like the rule; will see if I can persuade intellij to do this for me in future)

sdwilsh replied:
If you can figure that out, and it's something we can check into the project, we'd take a PR for that :)

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -61,7 +70,6 @@
 
   @Override
   public int execute(ExecutionContext context) throws IOException, InterruptedException {
-
```
nit: a tiny bit of a spurious change still :)

grumpyjames replied:
I'll fix the others in return for leaving this one ;-)

sdwilsh replied:
If you can figure out how to make that a project setting, we'd take a PR for that too :)

sdwilsh commented on src/com/facebook/buck/jvm/groovy/GroovycStep.java:
```diff
@@ -91,14 +98,63 @@ public String getDescription(ExecutionContext context) {
+          // compiled by groovyc).
+          if (!option.equals("sourcepath")) {
+            command.add("-J" + String.format("%s=%s", option, value));
+          }
```
This logic might be a tiny bit clearer if you just did a return in the `if` block and afterwards had the common case.

sdwilsh commented:
I'm otherwise happy with this, and can merge it in once you fix the lint error.

sdwilsh commented:
@facebook-github-bot shipit
